src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,213 @@
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use std::path::PathBuf;
+
+use error_chain::bail;
+
+use super::*;
+use crate::process;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let home = utils::cargo_home()?;
+        let cargo_bin = format!("{}/bin", cargo_home_str()?);
+        let env_name = home.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+// TODO: Update into a bytestring.
+pub fn cargo_home_str() -> Result<Cow<'static, str>> {
+    let path = utils::cargo_home()?;
+
+    let default_cargo_home = utils::home_dir()
+        .unwrap_or_else(|| PathBuf::from("."))
+        .join(".cargo");
+    Ok(if default_cargo_home == path {
+        "$HOME/.cargo".into()
+    } else {
+        match path.to_str() {
+            Some(p) => p.to_owned().into(),
+            None => bail!("Non-Unicode path!"),
+        }
+    })
+}
+
+// TODO: Tcsh (BSD)
+// TODO?: Make a decision on Ion Shell, Power Shell, Nushell
+// Cross-platform non-POSIX shells have not been assessed for integration yet
+fn enumerate_shells() -> Vec<Shell> {
+    vec![Box::new(Posix), Box::new(Bash), Box::new(Zsh)]
+}
+
+pub fn get_available_shells() -> impl Iterator<Item = Shell> {
+    enumerate_shells().into_iter().filter(|sh| sh.does_exist())
+}
+
+pub trait UnixShell {
+    // Detects if a shell "exists". Users have multiple shells, so an "eager"
+    // heuristic should be used, assuming shells exist if any traces do.
+    fn does_exist(&self) -> bool;
+
+    // Gives all rcfiles of a given shell that rustup is concerned with.
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
+
+    // Gives rcs that should be written to.
+    fn update_rcs(&self) -> Vec<PathBuf>;
+
+    // Writes the relevant env file.
+    fn env_script(&self) -> ShellScript {
+        ShellScript {
+            name: "env",
+            content: include_str!("env.sh"),
+        }
+    }
+
+    fn source_string(&self) -> Result<String> {
+        Ok(format!(r#"source "{}/env""#, cargo_home_str()?))
+    }
+}
+
+struct Posix;
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match utils::home_dir() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist. It's the only rc in the
+        // POSIX spec so it should always be set up.
+        self.rcfiles()
```

Comment:
With the suggestion above, rcfiles here could be inlined into update_rcs for this impl.

Reply:
unix.sh, line 65:
```rust
        for rc in sh.rcfiles().iter().filter(|rc| rc.is_file()) {
```

Reply:
I know it's basically that one place. ^^;
But it is a method that has a reason for existing.